File: social-network/python/tests/network_builder.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta, timezone

from social_network import Network

from .fixed_clock import FixedClock


@dataclass
class _SeededPost:
    user: str
    content: str
    minutes_after_start: int


@dataclass
class _SeededFollow:
    follower: str
    followee: str


class NetworkBuilder:
    def __init__(self) -> None:
        self._start_time = datetime(2026, 1, 15, 9, 0, 0, tzinfo=timezone.utc)
        self._seeded_posts: list[_SeededPost] = []
        self._seeded_follows: list[_SeededFollow] = []

    def starting_at(self, dt: datetime) -> NetworkBuilder:
        self._start_time = dt
        return self

    def with_post(self, user: str, content: str, minutes_after_start: int = 0) -> NetworkBuilder:
        self._seeded_posts.append(_SeededPost(user, content, minutes_after_start))
        return self

    def with_follow(self, follower: str, followee: str) -> NetworkBuilder:
        self._seeded_follows.append(_SeededFollow(follower, followee))
        return self

    def build(self) -> tuple[Network, FixedClock]:
        clock = FixedClock(self._start_time)
        network = Network(clock)

        for seed in self._seeded_posts:
            clock.advance_to(self._start_time + timedelta(minutes=seed.minutes_after_start))
            network.post(seed.user, seed.content)

        for seed in self._seeded_follows:
            network.follow(seed.follower, seed.followee)

        return network, clock

File: social-network/typescript/tests/NetworkBuilder.ts
import { Network } from '../src/Network.js';
import { FixedClock } from './FixedClock.js';

interface SeededPost {
  user: string;
  content: string;
  minutesAfterStart: number;
}

interface SeededFollow {
  follower: string;
  followee: string;
}

export class NetworkBuilder {
  private _startTime = new Date(Date.UTC(2026, 0, 15, 9, 0, 0));
  private _seededPosts: SeededPost[] = [];
  private _seededFollows: SeededFollow[] = [];

  startingAt(date: Date): this {
    this._startTime = date;
    return this;
  }

  withPost(user: string, content: string, minutesAfterStart = 0): this {
    this._seededPosts.push({ user, content, minutesAfterStart });
    return this;
  }

  withFollow(follower: string, followee: string): this {
    this._seededFollows.push({ follower, followee });
    return this;
  }

  build(): { network: Network; clock: FixedClock } {
    const clock = new FixedClock(this._startTime);
    const network = new Network(clock);

    for (const { user, content, minutesAfterStart } of this._seededPosts) {
      clock.advanceTo(new Date(this._startTime.getTime() + minutesAfterStart * 60_000));
      network.post(user, content);
    }

    for (const { follower, followee } of this._seededFollows) {
      network.follow(follower, followee);
    }

    return { network, clock };
  }
}
